engine/runtime-compiler/src/main/java/org/enso/compiler/common_logic/BuildScopeFromModuleAlgorithm.java

```diff
@@ -0,0 +1,191 @@
+package org.enso.compiler.common_logic;
+
+import org.enso.compiler.MetadataInteropHelpers;
+import org.enso.compiler.core.ir.Module;
+import org.enso.compiler.core.ir.module.scope.Definition;
+import org.enso.compiler.core.ir.module.scope.definition.Method;
+import org.enso.compiler.core.ir.module.scope.imports.Polyglot;
+import org.enso.compiler.data.BindingsMap;
+import org.enso.compiler.pass.resolve.MethodDefinitions$;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * Gathers the common logic for building the ModuleScope.
+ *
+ * <p>This is done in two places:
+ *
+ * <ol>
+ *   <li>in the compiler, gathering just the types to build StaticModuleScope,
+ *   <li>in the runtime, building Truffle nodes for the interpreter.
+ * </ol>
+ *
+ * <p>The interpreter does much more than the type-checker, so currently this only gathers the
+ * general shape of the process to try to ensure that they stay in sync. In future iterations, we
+ * may try to move more of the logic to this common place.
+ */
+public abstract class BuildScopeFromModuleAlgorithm<
+    FunctionType,
+    TypeScopeReferenceType,
+    ImportExportScopeType,
+    ModuleScopeType extends
+        CommonModuleScopeShape<FunctionType, TypeScopeReferenceType, ImportExportScopeType>,
+    ModuleScopeBuilderType extends
+        CommonModuleScopeShape.Builder<
+                FunctionType, TypeScopeReferenceType, ImportExportScopeType, ModuleScopeType>> {
+  private final Logger logger = LoggerFactory.getLogger(BuildScopeFromModuleAlgorithm.class);
+
+  /** The scope builder to which the algorithm will register the entities. */
+  protected final ModuleScopeBuilderType scopeBuilder;
```

Comment:
Why is this field `protected`?

Reply:
I think child classes were also using it in the implementations of the `protected` methods. If that ceased to be the case and I forgot to notice, I will make it `private`. Otherwise, if it's used in child classes, do you think it can stay `protected`?

Because alternatively, I could make it `private` and store yet another reference to the scope builder in each child class. But storing the same reference twice in the same class felt weird for me and I didn't think it was necessary, that's why I opted for `protected final`. But I'm open to a different approach if you think it's preferred.